File: client/src/features/vision-workspace/services/registry.ts
// registry.ts — despacho de estrategia por model_type. Unico lugar de despacho:
// agregar un tipo nuevo = crear <tipo>.service.ts + registrarlo aca.

import type { ModelType, VisionStrategy } from './types';
import { detectionStrategy } from './detection.service';
import { classificationStrategy } from './classification.service';
import { segmentationStrategy } from './segmentation.service';

export const VISION_STRATEGIES: Record<ModelType, VisionStrategy> = {
  detection: detectionStrategy,
  classification: classificationStrategy, // stub
  segmentation: segmentationStrategy, // stub
};

export function getStrategy(type: ModelType): VisionStrategy {
  return VISION_STRATEGIES[type];
}

File: src/api/func/tasks/registry.py
# tasks/registry.py — despacho por model_type.
# Reemplaza al viejo _PIPELINE_BUILDERS del controller: una sola fuente de verdad
# que mapea model_type -> TaskStrategy. Agregar un tipo nuevo es registrar su
# estrategia aca (un archivo nuevo en tasks/), sin tocar el controller ni el WS.

from .strategy import TaskStrategy
from .errors import UnknownModelType
from .detection import detection_strategy
from .classification import classification_strategy
from .segmentation import segmentation_strategy

TASK_STRATEGIES = {
    s.task: s
    for s in (detection_strategy, classification_strategy, segmentation_strategy)
}


def get_strategy(model_type: str) -> TaskStrategy:
    """Devuelve la TaskStrategy del model_type. Levanta UnknownModelType si no existe."""
    try:
        return TASK_STRATEGIES[model_type]
    except KeyError:
        raise UnknownModelType(
            f"model_type '{model_type}' no reconocido. "
            f"Validos: {sorted(TASK_STRATEGIES)}"
        ) from None
